src/statemachine/blender.js

```diff
@@ -0,0 +1,244 @@
+    }
+
+    function smooth(start, end, t) {
+        return interpolate.value(start, end, pc.math.smootherstep(0, 1, t));
+    }
+
+    /**
+     * @callback
+     * @name pc.interpolate~valuecb
+     * @description call back containing the value of an interpolation
+     * @param {pc.Vec3|pc.Quat|Number} value The current value
+     * @param {Number} t The current proportion of the interpolation that is complete 0..1
+     */
+
+    var interpolate = {
+        /**
+         * @name pc.interpolate.value
+         * @function
+         * @description Get a value between start and end with unconstrained t (can be > 1 or < 0)
+         * @param {pc.Vec3|pc.Quat|Number} start The start point for the interpolation
+         * @param {pc.Vec3|pc.Quat|Number}end The end point for the interpolation
+         * @param {Number} t The position between start and end
+         * @returns {pc.Vec3|pc.Quat|Number} The point at the specified time
+         */
+        value: value,
+        /**
+         * @name pc.interpolate.lerp
+         * @function
+         * @description lerp a value between start and end with constrained t (forced between 0 and
+         *     1)
+         * @param {pc.Vec3|pc.Quat|Number} start The start point for the interpolation
+         * @param {pc.Vec3|pc.Quat|Number} end The end point for the interpolation
+         * @param {Number} t The position between start and end
+         * @returns {pc.Vec3|pc.Quat|Number} The lerped value
+         */
+        lerp: lerp,
+        /**
+         * @name pc.interpolate.smooth
+         * @function
+         * @description Use smoothstep to make a natural lerp of a value between start and end with
+         *     constrained t (forced between 0 and 1)
+         * @param {pc.Vec3|pc.Quat|Number} start The start point for the interpolation
+         * @param {pc.Vec3|pc.Quat|Number} end The end point for the interpolation
+         * @param {Number} t The position between start and end
+         * @returns {pc.Vec3|pc.Quat|Number} The smoothly interpolated value
+         */
+        smooth: smooth,
+        /**
+         * @name pc.interpolate.moveTowards
+         * @function
+         * @description Move a value towards a target with a limited maximum step
+         * @param {pc.Vec3|pc.Quat|Number} start The start point for the move
+         * @param {pc.Vec3|pc.Quat|Number}end The target for the move
+         * @param {Number} max The maximum distance to move, if start and end are vectors or
+         *     numbers then this is a distance, if start and are quaternions then this is a
+         *     distance measured in radians
+         * @example
+         * update: function(dt) {
+		 *     this.entity.setPosition(pc.interpolate.moveTowards(this.entity.getPosition(),
+		 *     someTargetPosition, this.speed * dt));
+		 * }
+         * @returns {pc.Vec3|pc.Quat|Number} The updated value
+         */
+        moveTowards: function (start, end, max) {
```

Comment:
So I noticed this method and probably others too are making clones of vectors, quats etc. Since these are methods that are meant to be called every frame do you think maybe you could use scratch vectors/quats for temporary operations instead so that we minimize the memory footprint?